### PR TITLE
Write categories to UFO's openTypeCategories key

### PR DESCRIPTION
**Pull request by madig**
Closes https://github.com/googlefonts/glyphsLib/issues/791

* Use glyph category and subCategory info and anchor data to generate Opentype GDEF categories and write them to the UFO's `public.openTypeCategories` lib key
* For roundtripping purposes, write the Designspace's first UFO's `public.openTypeCategories` lib key into the font metadata and put it back in verbatim without guessing, but still generate categories for new glyphs. This means that when going to UFO and back, you cannot change the categories from Glyphs.app anymore, except by clearing out the font's `com.schriftgestaltung.Glyphs.originalOpenTypeCategory` userData key or changing them in the UFOs.

Todo:
- [x] Make new `--no-generate-GDEF` switch
- [x] Make `--generate-GDEF` switch a no-op, as in default behavior
- [x] When ufo2glyphs, store categories in userData to preserve it for roundtripping

### CODE REVIEW

**Comment by madig** [comment]
I am struggling to find the motivation to put openTypeCategories into a master's userData when ufo2glyphs-ing. If you want to avoid Glpyhs.app conventions from overwriting your carefully curated categories, maybe using `--no-generate-GDEF` during glyphs2ufo suffices?

**Comment by madig** [comment]
Not entirely sure how this interacts with regeneration of categories for bracket glyphs.

**Comment by anthrotype** [comment]
> Make new --no-generate-GDEF switch
> Make --generate-GDEF switch a no-op, as in default behavior

so this effectively changes the default behaviour of the glyphs2ufo command as regards GDEF data; previously --generate-GDEF was opt-in whereas it now --no-generate-GDEF is an opt-out. Are we ok with that? I wonder if that risks breaking somebody's workflow unexpectedly.

**Review by anthrotype — APPROVED** [review]

**Comment by madig** [comment]
> what do you mean exactly?

Bracket glyphs are generated on the fly and take on the category of their "parent". We have a test for that but I may be missing something.

> The regression tests are failing, I imagine that's expected and they would need to be re-generated, right?

Yes, expected, and no, regression tests are done by the workflow and need no regeneration I think.

> so this effectively changes the default behaviour of the glyphs2ufo command as regards GDEF data; previously --generate-GDEF was opt-in whereas it now --no-generate-GDEF is an opt-out. Are we ok with that? I wonder if that risks breaking somebody's workflow unexpectedly.

Probably. Before, we skipped stuff because we didn't want to roundtrip GDEF, but now there's no immedediately obvious harm in writing out categories.

File: Lib/glyphsLib/builder/features.py [hunk 1]
```diff
@@ -151,20 +155,14 @@ def _to_glyphs_language(langID):
 
 
 def _to_ufo_features(
-    font, ufo=None, generate_GDEF=False, skip_export_glyphs=None, master=None
-):
+    font: GSFont,
+    ufo: Font | None = None,
```
[the remark this answers is not in the document]
Comment (madig) [reply Lib/glyphsLib/builder/features.py:159]: Also, I think Font is close enough to defcon.Font that we shouldn't bother 😬

**Comment by anthrotype** [comment]
> now there's no immedediately obvious harm in writing out categories.

well, they will trigger ufo2ft's GdefFeatureWriter to create a GDEF GlyphClassDefs and LigatureCarets subtables, in theory one may not want that to happen as result of simply passing one's UFOs through glyphsLib and back to UFO.

**Comment by madig** [comment]
> well, they will trigger ufo2ft's GdefFeatureWriter to create a GDEF GlyphClassDefs and LigatureCarets subtables, in theory one may not want that to happen as result of simply passing one's UFOs through glyphsLib and back to UFO.

So, revert to "no" being the default?

**Comment by anthrotype** [comment]
> So, revert to "no" being the default?

isn't it already the case that the glyphs2ufo CLI tool has a different default than the API (UFOBuilder, etc.) as regards generation of GDEF? That could be at the same time, an argument for either unifying the defaults across CLI and API for consistency reasons, or for keeping the current behavior to reduce surprises.. I'm 50/50, you chose.

**Comment by madig** [comment]
K. Will turn it off by default to avoid (bigger) disruptions for now.

**Comment by anthrotype** [comment]
> Will turn it off by default to avoid (bigger) disruptions for now.

I'm also touching these files, so I'll merge and change that myself, thanks

**Pull request merged**